marimbabot_hardware/src/servo_interface.cpp

```diff
@@ -0,0 +1,199 @@
+        joint_state_interface.getHandle(servo_state.name), &servo_state.command);
+    position_joint_interface.registerHandle(position_servo_handle);
+
+    registerInterface(&position_joint_interface);
+
+    // Creating the joint saturation interface + handle
+    joint_limits_interface::getJointLimits(servo_state.name, node_handle, servo_limits);
+    joint_limits_interface::PositionJointSaturationHandle saturation_handle(
+        position_servo_handle, servo_limits);
+    position_joint_saturation_interface.registerHandle(saturation_handle);
+
+    registerInterface(&position_joint_saturation_interface);
+}
+
+void ServoInterface::initialize() {
+    // Keeping the current position of the servo
+    read();
+    servo_state.command = servo_state.position;
+    previous_command = servo_state.command;
+}
+
+bool ServoInterface::try_open_serial_port() {
+    try {
+        arduino_serial.open();
+        last_arduino_error = 0;
+        ROS_INFO_STREAM("Servo controller: Successfully opened port " << arduino_serial.getPort());
+        return true;
+    } catch (serial::IOException &e) {
+        if(e.getErrorNumber() != last_arduino_error) {
+            ROS_ERROR_STREAM("Servo controller error: Unable to open port " << arduino_serial.getPort());
+            ROS_ERROR_STREAM("Servo controller error: Error: " << e.what());
+            last_arduino_error = e.getErrorNumber();
+        }
+        return false;
+    }
+}
+
+void ServoInterface::read() {
+    // Calculating time since last update
+    ros::Time current_time = ros::Time::now();
+    last_run_period = current_time - last_run_time;
+    last_run_time = current_time;
+
+    // Checking if the arduino is connected
+    if(!arduino_serial.isOpen()) {
+        // Trying to reconnect
+        if(!try_open_serial_port()) {
+            return;
+        }
+    }
+
+    // Sending command to arduino
+    try {
+        arduino_serial.write("p\n");
+        arduino_serial.flush();
+    } catch (serial::SerialException &e) {
+        ROS_ERROR_STREAM("Servo controller error: Unable to write to port " << arduino_serial.getPort());
+        ROS_ERROR_STREAM("Servo controller error: Error: " << e.what());
+        arduino_serial.close();
+        return;
+    }
+
+    // Reading response from arduino
+    std::string response;
+    do {
+        try {
+            response += arduino_serial.readline();
+        } catch (serial::SerialException &e) {
+            ROS_ERROR_STREAM("Servo controller error: Unable to read from port " << arduino_serial.getPort());
+            ROS_ERROR_STREAM("Servo controller error: Error: " << e.what());
+            arduino_serial.close();
+            return;
+        }
+        //ROS_INFO_STREAM("Servo controller: Received response from arduino: \"" << response << "\"");
+    } while(arduino_serial.available() > 0);
+
+    // Parsing response (extract int value)
+    // remove first 2 chars and last chars (\r\n)
```

Comment:
I guess it makes sense to validate the first 2 chars if they are part of your protocol?
Or can drop them altogether?